File: flow_code/matlab/main_for_KITTI_complex.m
clear; close all; clc;

% pic_num = 409;
pic_num = 924;
path = "E:/Program Files/dataset/KITTI/2011_09_26_drive_0101_sync";
original_img_path = path+'/image_2/0000000'+num2str(pic_num)+'.png';
optical_flow_path = path+'/flow/0000000'+num2str(pic_num)+'.png';
semantic_path = path + '/semantic/0000000'+num2str(pic_num)+'.png';

% 图1 原始RGB
rgb_img = imread(original_img_path);
fig1 = figure;
imshow(rgb_img)
title("raw RGB")
axis off
addToolbarExplorationButtons(gcf)


% 图2 光流图

optical_flow_img = imread(optical_flow_path);

flow_u = (double(optical_flow_img(:,:,1))-2^15)/64.0;
flow_v = (double(optical_flow_img(:,:,2))-2^15)/64.0;
flow_valid = optical_flow_img(:,:,3);
flow_u(flow_valid==0) = nan;
flow_v(flow_valid==0) = nan;

fig2 = figure;
subplot(2,1,1); imshow(flow_u, [], 'Colormap', jet(4096));title("original fu")
subplot(2,1,2); imshow(flow_v, [], 'Colormap', jet(4096));title("original fv")

addToolbarExplorationButtons(gcf)


% 获取路面的mask
semantic_img = imread(semantic_path);
%imshow(semantic_img)
mask=zeros(size(semantic_img,1),size(semantic_img,2),"double");
for i = 1 : size(semantic_img,1)
    for j = 1 : size(semantic_img,2)
        if semantic_img(i,j,1)==255 ...
                && semantic_img(i,j,2)==0 ...
                && semantic_img(i,j,3)==0
            
            mask(i,j)=1;
        end
    end
end


%图3 光流fv图带掩码

flow_u = flow_u .* mask;
flow_v = flow_v .* mask;

figure;
imshow(flow_v, [], 'Colormap', jet(4096));title("original fu")
title("fv with mask")
axis('off')


% 图4 v-fv曲线图

[v_max, u_max, ~] = size(flow_u);
image_h = v_max;
image_w = u_max;

% pre-set parameter
u0 =610; %609.5593;
v0 =173; %172.8540;
fx = 721.5377;
fy = 721.5377;
f = 721.5377;
l = 2.7; %2500~2700 mm

% 选取mask所在点作为拟合的数据
[mask_v,mask_u] = find(mask~=0); % 非零的索引
fv_val=zeros(size(mask_v));
fu_val=zeros(size(mask_v));

for i = 1: size(mask_v,1)
    fv_val(i)=flow_v(mask_v(i), mask_u(i));
    fu_val(i)=flow_u(mask_v(i), mask_u(i));
end

mask_vu=horzcat(mask_v,mask_u);
% flow_val =vertcat(fv_val, fu_val);
% mask_vu=mask_vu';
% flow_val=flow_val';

% 曲线拟合
popt=lsqcurvefit('flow_KITTI',[0 0 0 0 1.5],mask_vu,[fv_val,fu_val],[-0.5 * pi  0  0 -0.5 * pi 1],[0.5 * pi 10 10 0.5 * pi 3]);
                    
theta_est = popt(1);
Xd_est= popt(2);
Zd_est = popt(3);
phi = popt(4);
h = popt(5);

fprintf("theta=%f,Xd=%f,Zd=%f,phi=%f,h=%f",theta_est*180/pi, Xd_est, Zd_est, phi*180/pi, h)

% 构造验证拟合效果的数组
u_valid=1:u_max;
v_valid=v0:v_max;
[v_square, u_square]=meshgrid(v_valid,u_valid);
u_array=reshape(u_square,[],1);
v_array=reshape(v_square,[],1);
valid_vu=double(horzcat(v_array,u_array));

flow_est=flow_KITTI(popt,valid_vu);
fv_est=flow_est(:,1);
fu_est=flow_est(:,2);
fv_est=reshape(fv_est,u_max,[]);
fv_est=fv_est'.*mask(v0:v_max,:);
fu_est=reshape(fu_est,u_max,[]);
fu_est=fu_est'.*mask(v0:v_max,:);

fig_estimation = figure;
ax = subplot(3,2,1); imshow(fv_est, [], 'Colormap', jet(4096));title('flow v estimation')
ax = subplot(3,2,2); imshow(fu_est, [], 'Colormap', jet(4096));title('flow u estimation')
ax = subplot(3,2,3); imshow(flow_v(v0:v_max,:), [], 'Colormap', jet(4096));title('flow v truth')
ax = subplot(3,2,4); imshow(flow_u(v0:v_max,:), [], 'Colormap', jet(4096));title('flow u truth')
ax = subplot(3,2,5); imshow(fv_est-flow_v(v0:v_max,:), [], 'Colormap', jet(4096));title('flow v error')
ax = subplot(3,2,6); imshow(fu_est-flow_u(v0:v_max,:), [], 'Colormap', jet(4096));title('flow v error')
axis off
addToolbarExplorationButtons(gcf)

hold on;
% plot(fv_tmp/dmax*100,V_tmp+oy2,'g*')
hold off;

% hold on Y=cfun(x); plot(Y, x, 'r','LineWidth',1)

File: flow_code/python/main_for_KITTI_complex.py
import cv2
import numpy as np
from PIL import Image
from matplotlib import pyplot as plt
from scipy.optimize import curve_fit
import math

# P_rect_03: f=721.5377 u0=6.095593 v0=172.8540

# get paths

# phi 比较大: 320~340 830~840 905~934
pic_num = 922
#v0 = 146  # 922
# v0 = 157  # 910
# v0 = 180  # 561
# pic_num = 116 330 331 410 411 532 560 561 699 700 875 876
path = "E:\Program Files\dataset\\KITTI\\2011_09_26_drive_0101_sync"
original_img_path = path + f"\\image_2\\0000000{pic_num}.png"
optical_flow_path = path + f"\\flow\\0000000{pic_num}.png"
semantic_path = path + f"\\semantic\\0000000{pic_num}.png"

# depth_path = path + r"\\depth\\00000040.png"

# ==================== 图1 原始RGB =========================================
rgb_img = Image.open(original_img_path)
plt.figure("raw RGB")
plt.imshow(rgb_img)
plt.axis('off')
plt.show()

original_img_path = path + f"\\image_2\\0000000{pic_num + 1}.png"
rgb_img = Image.open(original_img_path)
plt.figure("raw RGB next")
plt.imshow(rgb_img)
plt.axis('off')
plt.show()
# ==================== 获取路面的mask =========================================
semantic_img = cv2.imread(semantic_path)

# road: (0, 0, 255)
mask = cv2.inRange(semantic_img, (0, 0, 255), (0, 0, 255))
mask = mask / 255
mask.astype(int)

# ==================== 图2 光流图 =========================================
optical_flow_img = cv2.imread(optical_flow_path, -1)
# valid = optical_flow_img[:, :, 0]
fu = optical_flow_img[:, :, 2].astype(np.float32)
fv = optical_flow_img[:, :, 1].astype(np.float32)
fu = (fu - 2 ** 15) / 64.0
fv = (fv - 2 ** 15) / 64.0
# fu[valid == 0] = np.nan
# fv[valid == 0] = np.nan


plt.figure("fu fv")
plt.subplot(2, 1, 1)
plt.imshow(fu)
plt.title("original fu")
plt.axis('off')

plt.subplot(2, 1, 2)

plt.imshow(fv)
plt.axis('off')

# plt.subplot(2, 1, 2)
# plt.imshow(fv)
# plt.title("original fv")
# plt.axis('off')

# ==================== 图3 光流fv图带掩码 =========================================
fv_origin = fv
fu = fu * mask
fv = fv * mask

plt.figure("fv with mask")
plt.imshow(fv)
plt.axis('off')
plt.show()

# ==================== 图4 v-fv曲线图 =========================================
v_max = image_h = fu.shape[0]
u_max = image_w = fu.shape[1]
u0 = 609.5593
v0 = 172.8540

# plt.figure("fu_u0")
# plt.plot(range(v_max), fu[:, round(u0)], 'r')
# plt.show()

for i in range(v_max):
    for j in range(u_max):
        if fv[i, j] + i > v_max:
            fv[i, j] = 0

v_fv_map = np.zeros((v_max, round(fv.max()) + 1))
for i in range(v_max):
    for j in range(u_max):
        if round(fv[i, j]) != 0:
            v_fv_map[i, round(fv[i, j])] += 1

plt.figure("v-fv-map")
plt.imshow(v_fv_map)
plt.show()

# ==================== v-fv公式计算图 =========================================
# fv = Zd(v1-v0)/(hf/(v1-v0)-Zd)
# Y = Zd*X^2 / (hf - Zd*X)
# f = fx = fy = image_w / 2.0


v1 = np.arange(0, v_max, 0.1)
v1 = v1[v1 - v0 > 0]
# v1 = np.arange(250, v_max, 0.1)


# ==================== 曲线拟合 =========================================

# 选取中间非零点
v_head = v0
v_tail = v_max
for i in np.arange(round(v0) + 1, v_max):
    if np.argmax(v_fv_map[i, :]) != 0:
        v_head = i
        break
for i in np.arange(v_max - 1, v_head, -1):
    if np.argmax(v_fv_map[i, :]) != 0:
        v_tail = i + 1
        break

# original
# k1=cos(phi)*Zd-sin(phi)*Xd
# k2=h*sin(phi)
# k3=h*cos(phi)
# fv=(f*h*v)/(f*k3-k2*v-k1*v)-v

# physical meaning
# k1=Zd-tan(phi)*Xd
# k2=phi
# k3=h
# fv=(f*h*v)/(f*h*cos(k2)-h*sin(k2)-k1*v*cos(k2))-v

vu = []
v_x = []
u_x = []
fv_val = []
for i in np.arange(v_head, v_tail):
    for j in np.arange(u_max):
        if fv[i, j] > 0:
            v_x.append(i)
            u_x.append(j)
            fv_val.append(fv[i, j])
# vu = np.append(v_x,u_x)
fv_val = np.array(fv_val)
vu = np.empty((2, fv_val.shape[0]), dtype="int32")
vu[0] = v_x
vu[1] = u_x
fv_val = fv_val[0:(v_tail - v_head) * u_max]

f = 958


# u = np.arange(u_max)
# v = np.arange(v_head, v_tail)
# x = np.append(u, v, axis=0)
# U, V = np.meshgrid(u, v)
# VV = np.expand_dims(V, 0)
# UU = np.expand_dims(U, 0)
# vu = np.append(VV, UU, axis=0)
# fv_val = fv.ravel()
# fv_val = fv_val[0:(v_tail - v_head) * u_max]


def fitting_curve(x, k1, k2, h):
    v = x[0]
    u = x[1]
    # fv = var1(v1 - v0) / (var2 / (v1 - v0) - var1)
    # output = var1 * (x - v0) ** 2 / (var2 - (x - v0))
    # output = var1 * (x - var3) ** 2 / (var2 - (x - var3))
    output = (f * h * (v - v0)) / \
             (f * h * math.cos(k2) - h * (u - u0) * math.sin(k2) - k1 * (v - v0) * math.cos(k2)) - (v - v0)

    return output
    #return output.ravel()

popt, _ = curve_fit(fitting_curve, vu, fv_val
                    , p0=[0.5, -0.2, 1.5], bounds=([0, -np.pi, 0], np.array([np.inf, 0, 2.3])))

# popt, _ = curve_fit(fitting_curve, np.arange(250, 300), np.argmax(v_fv_map, 1)[250:300],
#                     p0=[1000, 200], bounds=(np.zeros(2), np.array([5000, 600])))

# p0=[0, 1000, 200], bounds=(np.zeros(3), np.array([10, 5000, 600])))
# Y = popt[0] * (v1 - popt[2]) ** 2 / (popt[1] - (v1 - popt[2]))
# fv_est = np.zeros_like(fv)

# 构造验证拟合效果的数组
u = np.arange(u_max)
v = np.arange(v_head, v_tail)
# x = np.append(u, v, axis=0)
U, V = np.meshgrid(u, v)
VV = np.expand_dims(V, 0)
UU = np.expand_dims(U, 0)
vu1 = np.append(VV, UU, axis=0)

# parameter estimation
k1_est = popt[0]
k2_est = popt[1]
h_est = popt[2]
# u0_est = popt[3]

# 拟合得到的fv
fv_est = fitting_curve(vu1, k1_est, k2_est, h_est)
fv_est = fv_est.reshape(v_tail - v_head, u_max)
# fv 误差
fv_diff = fv_est - fv_origin[v_head:v_tail]
# for x in vu:
#     fv_est[x[0], x[1]] = (f * h_est * (x[0] - v0)) / \
#                          (f * h_est * math.cos(k2_est) - h_est * (x[1] - u0) * math.sin(k2_est) - k1_est * (
#                                  x[0] - v0) * math.cos(k2_est)) - (x[0] - v0)
# Y = (v1 - popt[1]) ** 2 / (popt[0] - (v1 - popt[1]))
plt.figure("fv estimate")
plt.subplot(3, 1, 1)
plt.imshow(fv_est)
plt.subplot(3, 1, 2)
plt.imshow(fv[v_head:v_tail, 0:u_max])
plt.subplot(3, 1, 3)
plt.imshow(fv_diff[:, 0:u_max])
# print(f"popt[0]:{popt[0]}, popt[1]:{popt[1]}, popt[2]:{popt[2]}")
print(f"k1:{k1_est}, k2:{k2_est}, h:{h_est}, phi:{k2_est*180/3.14}")
# plt.plot(np.argmax(v_fv_map, 1)[250:], np.arange(250, v_max), '-b')

# 通过特殊点估计参数
# ftan = fu[round(v0), round(u0)]
# print(f"f*tan(phi)={ftan:.6f}")
